components/core/src/clp/SysFileReader.cpp

```diff
@@ -0,0 +1,66 @@
+#include "SysFileReader.hpp"
+
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <cstddef>
+#include <cstdio>
+
+#include "ErrorCode.hpp"
+
+namespace clp {
+auto SysFileReader::try_read(char* buf, size_t num_bytes_to_read, size_t& num_bytes_read)
+        -> ErrorCode {
+    if (nullptr == buf) {
+        return ErrorCode_BadParam;
+    }
+
+    num_bytes_read = 0;
+    while (true) {
+        auto const bytes_read = ::read(m_fd.get_raw_fd(), buf, num_bytes_to_read);
+        if (0 == bytes_read) {
+            break;
+        }
+        if (bytes_read < 0) {
+            return ErrorCode_errno;
+        }
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        buf += bytes_read;
```

Comment:
We should use `std::span` to wrap the buffer instead. Maybe check `NetworkReader` and I had implemented similar logic.